### ISSUE TITLE
Bedrock edition - empty sub-chunks not created

### ISSUE THREAD

**Issue by gentlegiantJGC** [Khroki/MCEdit-Unified#824]
When filling or modifying the world in any way, sub-chunks that do not previously have any blocks in them do not save properly. I do not know if they are written but incorrectly or not written at all but when the world is refreshed or loaded in game any sub-chunk that did not previously have any blocks in it will still not have any blocks in it.
a sub-chunk being a 16x16x16 region within a chunk if that is not standard nomenclature

Running from source on windows

**Comment by gentlegiantJGC** [comment]
So are you saying that if a sub-chunks is modified in game all sub-chunks below it (and including it I assume) are saved?
Does this mean that a workaround would be to place blocks in each chunk in the top sub-chunk

**Comment by LaChal** [comment]
I may be wrong, but it's basically the behaviour of PE engine.
Putting blocks in the highest position in every chunk you want to edit can be tedious, but unless Mojang gives something to create chunks, I'm afraid it's the sole way...

**Comment by gentlegiantJGC** [comment]
I won't have a chance to test this for a while but I will assume it works.

My main issue was that when a schematic is pasted into sub-chunks that do not exist MCedit will show the blocks in the non-existent sub-chunks but after saving and reloading they are not there because the sub-chunks could not be created. Is there a way to show the user that those blocks will not be saved? Eg highliting the blocks in that sub-chunk in some way / not rendering them in the first place

**Comment by LaChal** [comment]
Not yet...

The only way for now may be to switch to chunk view and navigate to chunks you suspect to not exists in game; or look to your pasted schematics from top in world view...

We had some discussion about that when we were aware of PE game is not storing unmodified chunks, and could not find a satisfying solution, except not saving data in ingame unmodified chunks.
Technically it is possible to generate chunks in MCEdit, but without an engine to create the terrain (and entities) in, they'll be filled up with air, leading in 'holes' ingame' :/
I guess it's not wanted...

Anyway, colouring such blocks/entities is the best option :smile:

**Comment by gentlegiantJGC** [comment]
My issue here isn't pasting into whole chunks that don't exist (but I suppose that would also be an issue) but pasting into sub-chunks that are just air above sub-chunks that have already been generated.
I may look into making a local modification that allows this

**Comment by LaChal** [comment]
We did (yet) not explored this on our side.
Our main goal was to provide something able to load, render and save PE worlds.
So, if you want to try to create the needed code to add sub-chunks above last existing one, feel free :)
Just fork this repo, make your changes and create a pull request. It will be reviwed, and maybe merged...

@gentlegiantJGC @neonerz I'm about to commit changes in a dozen of source files to uniform the calls to some internal definition files handling.
To perform tests on my side, I updated the `mcver/PE/blocks.json` and `terrain-pocket.png` files from
[#820](https://github.com/Khroki/MCEdit-Unified/issues/820#issuecomment-312742221)
Do you agree these modified files can be pushed on this repository?

**Comment by neonerz** [comment]
@LaChal The blocks.json and terrain-pocket.png I have loaded on my github (linked in #820) should be good to go. I'm missing bone blocks from 1.3 in blocks.json though. Had a bit of trouble adding them, and haven't had a chance to fix it.

**Comment by neonerz** [comment]
@gentlegiantJGC One of the biggest issues with the conversion filter I made for converting Java to PE (besides #800) is the subchunk issue. If you are able to get it resolved in your fork, let me know. Even if it doesn't get merged into mcedit proper, that should be all I need to make my filter usable to the average person and will throw it up on my github.

**Comment by gentlegiantJGC** [comment]
Ok. When I get back off holiday I will try and find some time to see if I can to that. No promises.

I didn't have any problems copying a schematic out of PE and into PC. Tried the other way around and it looked alright but I didn't actually place or save it

**Comment by gentlegiantJGC** [comment]
@LaChal @neonerz 
Done a lot of diggin and found that MCedit will in fact write sub-chunks to disk that have not been created in game. The issue is is that the sub-chunk is not created when MCedit loads the world. The code to do so is fairly easy and I have created a filter that will create sub-chunks in a selection.

I haven't looked to far into the sub-chunk version tag but the sub-chunk I extracted from game had a tag of 0 so I hard coded that in the filter. If that is wrong I can change it.

As for whole chunks that are not created in the first place. This may be possible to do by creating the chunk in MCedit however like LaChal said there won't be any terrain in it and I don't know how the game will deal with the empty chunk.

The filter:
https://github.com/gentlegiantJGC/MCEdit-Filters/blob/master/Bedrock%20Edition%20Filters/createSubChunks.py

**Comment by LaChal** [comment]
The logic looks good to me :smile:

The sub-chunk version is used to calculate (or not) the chunk lights in PE 1+ chunks.
Valid values are, for now '\x03' and '\x04'.
PE 1+ engine will or not regenerate lights, so with the later version of chunks, it's not necessary to save MCEdit computed lights, nor to read these values from the DB data.

Basically, the sub-chunk data used in MCEdit is 10241 bytes long (see [line 1785 in leveldbpocket.py](https://github.com/Khroki/MCEdit-Unified/blob/master/pymclevel/leveldbpocket.py#L1785)):
* 1 byte: sub-chunk version.
* 4096 bytes for blocks.
* 2048 bytes for data.
* 2048 bytes for sky light.
* 2048 bytes for blocks light.

If the sub-chunk version is '\x04', only the first 6045 bytes are used (version, blocks and data).

Since your code does not create the whole chunk, I guess you can get the chunk version and use it as the sub-chunk version.
Also, if the chunk version is '\x02', just bypass, since pre 1.0 chunks are monolithic.

Checking the 'level.gameVersion' is also a good idea, because pre 1.0 worlds can't contain 1+ chunks.

The empty chunk will be simply full of air, and not re-generated by the game, I guess.

P.S.: You can create a list of faulty chunk to be populated in the 'except' statement, then use the 'alert' function to display the message in the UI like in the 'Find' filter:smile:

**Comment by gentlegiantJGC** [comment]
Thanks. I was just basing it on a chunk in the world I had open that was only 6045 bytes long.

When I tried removing the ground while in game on a superflat world and relogged it regenerated the empty chunk which is why I wasn't sure how it would behave.

I have done custom popup boxes before but it might get a bit much if a lot are selected and it is probably unneccarly complicated.

So what values of level.gameVersion should raise an exception? and should the version tag be hard coded as 4 or get the chunk version and use that?

**Comment by LaChal** [comment]
The 'level.gameVersion' you shall bypass id '\x02' (actually every version which is not PE 1+).

To get the chunk version just use 'chunk.version'. I would prefer this option since creating '\x04' sub-chunks in '\x03' chunks can lead to data corruption in the game. (Not sure...)

> When I tried removing the ground while in game on a superflat world and relogged it regenerated the empty chunk which is why I wasn't sure how it would behave.

Do you mean you dug all chunk terrain ingame and when you logged again in the game it was regenerated?
Or, you was logged in the game, removed the chunk in MCEdit, then relogged in the game?

> I have done custom popup boxes before but it might get a bit much if a lot are selected and it is probably unneccarly complicated.

Right, displaying all the faulty chunks in a dialog box can be an annoyance.
But using a box to tell people to have a look to the console before closing MCEdit may be useful.
Another option is to put the console messages in the log file too.
(Not really important yet. Having working, non crashing and non data corrupting stuff is more important :wink:)

**Comment by gentlegiantJGC** [comment]
Ok I will make those changes at some point.

All in game. I hadn't got MCedit working at that point. I used the fill command to remove everything and it regenerated it on relog.

I will probably get it to do a popup using raise exception at the end like texelelf does in his nbt editor filter

**Comment by gentlegiantJGC** [comment]
@LaChal is there a safe way to directly add values to the leveldb database? I have currently got the following code in a filter to write the chunk version tag to the database (I believe this is what is necessary for the game to recognise a chunk)

```
level.createChunk(cx,cz)
chunk = level.getChunk(cx, cz)
with level.worldFile.world_db() as db:
	key = struct.pack('<i', cx) + struct.pack('<i', cz)
	wop = level.worldFile.writeOptions
	db.Put(wop, key + '\x76', chunk.version)
```

This bypasses the save functionality of MCedit and writes that key to the database whether you save or not. The issue is if the world gets reloaded without saving, the version tag gets written but the block array doesn't because that is done when the player saves. It then requires the player to run the filter again, save, reload and then edit the chunk. If the world wasn't reloaded without saving the first time the chunk could be edited without reloading.

TLDR: is there a way to add values to the database that get written when the user saves the world rather than when the filter is run?

**Comment by LaChal** [comment]
You're using the C/C++ to Python interface directly, which means, as you remarked, that MCEdit is not aware of the changes you made there. It may lead to to data corruption or weird stuff.

When calling `level.createChunk`, the created chunk is empty. No entities nor blocks, PE 1+ chunks don't have any sub-chunk.
When saving the world in MCEdit, only the existing sub-chunks of a PE 1+ one are saved on the disk. ~~But the version will be written.~~ (*)

By default `PocketLeveldbChunk1Plus` is instantiated with `version` set to '\x03'. The attribute is not protected for writing, so you can change it after the chunk is created. Just pay attention to not change this if the chunk is not empty!

When the filter `perform` method is called (when clicking the `Filter` button in the GUI), MCEdit records the changes.

(*) I've just remarked that nothing currently writes the version back to the database.
Will be added :smile:

**Comment by gentlegiantJGC** [comment]
ok great thanks. That should simplify the filter a bit.
So there is no safe way to access the database like you would with NBT on the PC version so that MCedit saves the changes when the user saves?

I saw there were keys relating to maps and would like to update the MapIt filter for bedrock. Would I just have to directly edit the database for those files like I was doing here?
